### DISCUSSION TITLE
String repository clearing

### DISCUSSION THREAD

**Discussion by momomomo888** [B3nedikt/restring#122] [Q&A] [answered]
Is there any way to remove all keys and locales from a repository (memory and persistent/sharedprefs)?

I'm using the default one, and I tried using Restring.getStringRepository().getStrings().clear() without success.

Thanks in advance!

**Comment by B3nedikt** [comment, marked as answer]
This test does check exactly that:

https://github.com/B3nedikt/restring/blob/5f153babc2de833ae8fb0c66b54a9fa66de738f4/restring/src/test/java/dev/b3nedikt/restring/repository/CachedStringRepositoryTest.kt#L49-L58

So I am not sure what is not working for you. Can you clone the repository and either modify the sample app or the test cases to reproduce your issue?

**Comment by momomomo888** [comment]
Sorry, you are right. I've got confused by after seeing the xml files being present at the app's sharedprefs data folder. But if I download the file, the map is empty.

<img width="825" alt="image" src="https://user-images.githubusercontent.com/7397774/222727273-c2420568-ad18-4143-911f-8fab52d752da.png">

Thank you very much, this library is excellent :)

**Reply by B3nedikt** [reply]
Thanks, glad you like it :)

**Comment by zhangchf** [comment]
Looks like the api is not exposed to user, like `Restring.stringRepository.strings.xxx()`.

I also need a way to clear all the caches, or the cache for a Locale, but didn't find a way to do that.

I also tried this, but doesn't work:
```
      (Restring.stringRepository as? CachedStringRepository)?.apply {
        strings.clear()
        quantityStrings.clear()
      }

**Comment by zhangchf** [comment]
Looks like `strings.clear()` works, but `quantityStrings.clear()` doesn't work.

My test code is like this:
```
  override fun onCreate() {
    super.onCreate()

    Restring.init(this)
    ViewPump.init(RewordInterceptor)

    CoroutineScope(Dispatchers.IO).launch {
      delay(3000)

      Restring.putStrings(localeEN, mapOf(
        "signUp" to "Sign Up (test $counter)",
      ))
      Restring.putQuantityStrings(localeEN, mapOf(
        "testPlural" to mapOf<PluralKeyword, CharSequence>(
          PluralKeyword.OTHER to "%1\$d test items"
        )
      ))
    }
```
    
Then I clear it like this:
```
override fun onCreate() {
    super.onCreate()

    Restring.init(this)
    ViewPump.init(RewordInterceptor)


    CoroutineScope(Dispatchers.IO).launch {
      delay(3000)
      val localeEN = Locale("en")
      (Restring.stringRepository as? CachedStringRepository)?.apply {
        strings.clear()
        quantityStrings.clear()
      }
    }
```

**Comment by zhangchf** [comment]
If I clear a specific Locale, it can work successfully.
```
      val localeEN = Locale("en")
      (Restring.stringRepository as? CachedStringRepository)?.apply {
        strings.remove(localeEN)
        quantityStrings.remove(localeEN)
      }
```

**Comment by B3nedikt** [comment]
The intended way to use this API would be:

```kotlin
        val localeEN = Locale("en")
        Restring.stringRepository.toMutableRepository().apply {
            strings.remove(localeEN)
            quantityStrings.remove(localeEN)
        }
```

I am not sure why it is not working for you to clear all Locales at once though. I quickly tried the following:

```kotlin
@RunWith(RobolectricTestRunner::class)
class CachedStringRepositoryTest {

...

    @Test
    fun shouldGetNoLocalesAndQuantityStringAfterCallingClear() {
        val quantityStrings = generateQuantityStrings(10)
        val strings = generateStrings(10)

        stringRepository.strings[Locale.ENGLISH]?.putAll(strings)
        stringRepository.strings[Locale.FRENCH]?.putAll(strings)

        stringRepository.quantityStrings[Locale.ENGLISH]?.putAll(quantityStrings)
        stringRepository.quantityStrings[Locale.FRENCH]?.putAll(quantityStrings)

        stringRepository.strings.clear()
        stringRepository.quantityStrings.clear()

        stringRepository.supportedLocales.size shouldBeEqualTo 0
        stringRepository.quantityStrings.size shouldBeEqualTo 0
        stringRepository.strings.size shouldBeEqualTo 0
    }

}
```

It passes for me. If you think there is a bug, try to modify this test case so that it no longer passes. Otherwise I do not how how to help, as I do not know how you integrated your test code. You can also try to modify the example app in this repository, so that it reproduces your issue.

I hope this helps ;)

**Comment by zhangchf** [comment]
Hi Benedikt, 

Thanks for the quick reply! I don't know what's happening, from the code, it looks like everything is alright.

I set up a minimal demo project which is attached (all the setup code is in `MyApplication.kt`). Here is the reproduce steps:
1. Run the app in `English` language, then kill the app.
2. Check and download related `shared_prefs` from `Device Explorer` (If you use an Emulator, then it should be `/data/data/com.zcf.testrstring/shared_prefs`). The `xxx_en.xml files` now has the data.
3. Change system language to `Deustch`, and launch the app again.
4. Check and download related `shared_prefs` in Device Explorer again. Now the data in those `xxx_en.xml files` should have been cleared, but they're not.

[TestRstring.zip](https://github.com/user-attachments/files/19626588/TestRstring.zip)
